File: 백준 문제 풀이 모음/class 4/1504.py
import sys, heapq
input = sys.stdin.readline
INF = float('inf')
N, E = map(int, input().split())
graph = [[] for i in range(N + 1)] 

for i in range(E) :
    a, b, c = map(int, input().split())
    graph[a].append((c, b))
    graph[b].append((c, a))

start, end = map(int, input().split())


def dijkstra(start) :
    distance = [INF ] * (N+1)
    distance[start] = 0
    heap = []
    heapq.heappush(heap, (0, start))

    while heap :
        now_cost, now_node = heapq.heappop(heap)
        
        if distance[now_node] < now_cost :
            continue

        for new_cost, new_node in graph[now_node] :
            new_cost = new_cost + now_cost
            if distance[new_node] > new_cost :
                distance[new_node] = new_cost
                heapq.heappush(heap, (new_cost, new_node))
        
    return distance

arr1 = dijkstra(1)
arr2 = dijkstra(start)
arr3 = dijkstra(end)

first_arr = arr1[start] + arr2[end] + arr3[N]
second_arr = arr1[end] + arr3[start] + arr2[N]
min_arr = min(first_arr, second_arr)
if min_arr >= INF :
    print(-1)
else :
    print(min_arr)

File: 백준 문제 풀이 모음/kotlin/class 4/1504.kt
import java.util.PriorityQueue

private const val INF = 200000001
private lateinit var list : MutableList<MutableList<Node>>

private class Node (var nodeNum :Int, var weight : Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return weight - other.weight
    }
}

fun main() = with(System.`in`.bufferedReader()){
    val input = readLine().split(' ').toList()
    val n = input[0].toInt()
    val e = input[1].toInt()

    list = MutableList<MutableList<Node>>(n+1) { MutableList<Node>(0) { Node(0,0)} }
    repeat(e) {
        val input = readLine().split(' ').toList().map { it.toInt() }
        list[input[0]].add(Node(input[1], input[2]))
        list[input[1]].add(Node(input[0], input[2]))
    }

    val input1 = readLine().split(' ').toList().map { it.toInt() }
    val v1 = input1[0]
    val v2 = input1[1]

    val arr1 = dikjstra(1, n)
    val arr2 = dikjstra(v1, n)
    val arr3 = dikjstra(v2, n)
//
//    println(arr1)
//    println(arr2)
//    println(arr3)

    val v1_cost = arr1[v1] +arr2[v2] + arr3[n]
    val v2_cost = arr1[v2] +arr3[v1] + arr2[n]

    val cost = Math.min(v1_cost, v2_cost)
    if (cost >= INF) {
        println(-1)
    } else {

        println(cost)
    }
}

fun dikjstra(start: Int, n: Int): MutableList<Int> {
    val dist = MutableList<Int>(n+1) { INF }
    dist[start] = 0
    val pq = PriorityQueue<Node>()
    pq.add(Node(start, 0))

    while (pq.isNotEmpty()) {
        val node = pq.poll()

        if (dist[node.nodeNum] < node.weight) continue

        list[node.nodeNum].forEach {
            if ( dist[it.nodeNum] > dist[node.nodeNum] + it.weight) {
                    dist[it.nodeNum] = dist[node.nodeNum] + it.weight
                    pq.add(Node(it.nodeNum, it.weight))
                }
            }
        }


    return dist
}
